pkg/storegateway/bucket.go

```diff
@@ -615,102 +620,269 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
+	if err := msg.Encode(srv, storepb.NewStreamSeriesResponse(seriesBatch)); err != nil {
+		return status.Error(codes.Internal, errors.Wrap(err, "encode streaming series response").Error())
+	}
+	// Send the message.
+	if err := srv.SendMsg(msg); err != nil {
+		return status.Error(codes.Unknown, errors.Wrap(err, "send streaming series response").Error())
+	}
+
+	if seriesSet.Err() != nil {
+		return errors.Wrap(seriesSet.Err(), "expand series set")
+	}
+
+	return nil
+}
+
+func (s *BucketStore) sendSeriesChunks(
+	req *storepb.SeriesRequest,
+	srv storepb.Store_SeriesServer,
+	seriesSet storepb.SeriesSet,
+	stats *safeQueryStats,
+) (err error) {
+	var (
+		iterationBegin  = time.Now()
+		encodeDuration  = time.Duration(0)
+		sendDuration    = time.Duration(0)
+		seriesCount     int
+		chunksCount     int
+		streamingChunks = req.StreamingChunksBatchSize > 0 && !req.SkipChunks
+	)
 
-		err = nil
+	// Once the iteration is done we will update the stats.
+	defer stats.update(func(stats *queryStats) {
+		stats.mergedSeriesCount += seriesCount
+		stats.mergedChunksCount += chunksCount
+
+		// The time spent iterating over the series set is the
+		// actual time spent fetching series and chunks, encoding and sending them to the client.
+		// We split the timings to have a better view over how time is spent.
+		stats.streamingSeriesFetchSeriesAndChunksDuration += stats.streamingSeriesWaitBatchLoadedDuration
+		stats.streamingSeriesEncodeResponseDuration += encodeDuration
+		stats.streamingSeriesSendResponseDuration += sendDuration
+		stats.streamingSeriesOtherDuration += time.Duration(util_math.Max(0, int64(time.Since(iterationBegin)-
+			stats.streamingSeriesFetchSeriesAndChunksDuration-encodeDuration-sendDuration)))
 	})
 
-	if err != nil {
-		return
+	var batchSizeBytes int
+	var chunksBuffer []*storepb.StreamSeriesChunks
+	if streamingChunks {
+		chunksBuffer = make([]*storepb.StreamSeriesChunks, req.StreamingChunksBatchSize)
+		for i := range chunksBuffer {
+			chunksBuffer[i] = &storepb.StreamSeriesChunks{}
+		}
+	}
+	chunksBatch := &storepb.StreamSeriesChunksBatch{
+		Series: chunksBuffer[:0],
+	}
+	for seriesSet.Next() {
+		// IMPORTANT: do not retain the memory returned by seriesSet.At() beyond this loop cycle
+		// because the subsequent call to seriesSet.Next() may release it.
+		lset, chks := seriesSet.At()
+		seriesCount++
+		var response interface{}
+		if streamingChunks {
+			// We only need to stream chunks here because the series labels have already
+			// been sent above.
+			chunksBatch.Series = chunksBatch.Series[:len(chunksBatch.Series)+1]
+			last := chunksBatch.Series[len(chunksBatch.Series)-1]
+			last.Chunks = chks
+			last.SeriesIndex = uint64(seriesCount - 1)
+
+			batchSizeBytes += last.Size()
+			if (batchSizeBytes > 0 && batchSizeBytes > queryStreamBatchMessageSize) || len(chunksBatch.Series) >= int(req.StreamingChunksBatchSize) {
+				response = storepb.NewStreamSeriesChunksResponse(chunksBatch)
+			}
```

Comment:
I copy pasted the logic from the ingesters. The main aim is to send a reasonably large request without hitting any limits. 1MiB limit here is the target size by not a strict requirement since even with a little overhead we will be under the limits (grpc, TCP).

> Can you document the reason for batching in the first place

Where would be the best place to document this? Since it is defined in protos, would that be the place? It is largely to reduce CPU overheads.

Reply:
I think documenting this on the proto file sounds good enough.